File: apps/ios/UNES/Features/Messages/Components/OriginSwatch.swift
import SwiftUI

/// Origin swatch shown at the left of every message row and on the detail
/// sender card. Shape and glyph depend on `message.origin`:
///
/// * `.discipline` — rounded rect in the discipline's accent color with the
///   short code on top.
/// * `.direct`     — circle with a subtle tinted ring + person glyph.
/// * `.secretariat`/`.campus`/`.app`/`.module` — rounded tile with tinted
///   background and a kind-specific icon.
struct OriginSwatch: View {
    let message: Message
    var size: CGFloat = 40

    var body: some View {
        let meta = message.meta
        switch message.origin {
        case .direct:
            directSwatch(meta: meta)
        case .discipline:
            disciplineSwatch(meta: meta)
        default:
            iconSwatch(meta: meta)
        }
    }

    private func directSwatch(meta: MessageOriginMeta) -> some View {
        ZStack {
            Circle()
                .fill(meta.color.opacity(0.09))
            Circle()
                .stroke(meta.color.opacity(0.4), lineWidth: 1.5)
            PersonGlyph()
                .stroke(meta.color, style: StrokeStyle(lineWidth: 1.3, lineCap: .round, lineJoin: .round))
                .frame(width: size * 0.38, height: size * 0.38)
        }
        .frame(width: size, height: size)
    }

    private func disciplineSwatch(meta: MessageOriginMeta) -> some View {
        ZStack {
            RoundedRectangle(cornerRadius: 11, style: .continuous)
                .fill(meta.color)
            Text(meta.label)
                .font(UNESFont.mono(10, weight: .bold))
                .tracking(0.4)
                .foregroundStyle(Color.white)
        }
        .frame(width: size, height: size)
    }

    private func iconSwatch(meta: MessageOriginMeta) -> some View {
        ZStack {
            RoundedRectangle(cornerRadius: 11, style: .continuous)
                .fill(meta.color.opacity(0.12))
            iconContent(for: message.origin)
                .foregroundStyle(meta.color)
        }
        .frame(width: size, height: size)
    }

    @ViewBuilder
    private func iconContent(for origin: MessageOrigin) -> some View {
        switch origin {
        case .app:
            // UNES speech card
            AppGlyph()
                .stroke(style: StrokeStyle(lineWidth: 1.4, lineJoin: .round))
                .frame(width: 18, height: 18)
        case .module:
            // 2×2 grid of tiles
            ModuleGlyph()
                .stroke(style: StrokeStyle(lineWidth: 1.4))
                .frame(width: 18, height: 18)
        case .secretariat:
            // envelope
            EnvelopeGlyph()
                .stroke(style: StrokeStyle(lineWidth: 1.3, lineCap: .round))
                .frame(width: 18, height: 18)
        case .campus:
            // graduation cap
            CampusGlyph()
                .stroke(style: StrokeStyle(lineWidth: 1.3, lineCap: .round, lineJoin: .round))
                .frame(width: 18, height: 18)
        default:
            EmptyView()
        }
    }
}

// MARK: - Glyphs (shape primitives scaled to the view's frame)

private struct PersonGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let sx = rect.width / 15.0
        let sy = rect.height / 15.0
        // head
        p.addEllipse(in: CGRect(x: (7.5 - 2.3) * sx, y: (5 - 2.3) * sy,
                                width: 4.6 * sx, height: 4.6 * sy))
        // shoulders
        p.move(to: CGPoint(x: 2.5 * sx, y: 13 * sy))
        p.addQuadCurve(to: CGPoint(x: 12.5 * sx, y: 13 * sy),
                       control: CGPoint(x: 7.5 * sx, y: 7.3 * sy))
        return p
    }
}

private struct AppGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        // chat card: M4 4h10v7l-4-2-4 3V4z
        var p = Path()
        let sx = rect.width / 18.0
        let sy = rect.height / 18.0
        p.move(to: CGPoint(x: 4 * sx, y: 4 * sy))
        p.addLine(to: CGPoint(x: 14 * sx, y: 4 * sy))
        p.addLine(to: CGPoint(x: 14 * sx, y: 11 * sy))
        p.addLine(to: CGPoint(x: 10 * sx, y: 9 * sy))
        p.addLine(to: CGPoint(x: 6 * sx, y: 12 * sy))
        p.closeSubpath()
        return p
    }
}

private struct ModuleGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let sx = rect.width / 18.0
        let sy = rect.height / 18.0
        let r: CGFloat = 1.2
        let side: CGFloat = 5.5
        for (x, y) in [(3, 3), (9.5, 3), (3, 9.5), (9.5, 9.5)] {
            p.addRoundedRect(
                in: CGRect(x: CGFloat(x) * sx, y: CGFloat(y) * sy,
                           width: side * sx, height: side * sy),
                cornerSize: CGSize(width: r * sx, height: r * sy)
            )
        }
        return p
    }
}

private struct EnvelopeGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let sx = rect.width / 18.0
        let sy = rect.height / 18.0
        // outer rect
        p.addRoundedRect(
            in: CGRect(x: 3 * sx, y: 4 * sy, width: 12 * sx, height: 10 * sy),
            cornerSize: CGSize(width: 1.4 * sx, height: 1.4 * sy)
        )
        // flap: M3 4l6 4.5L15 4
        p.move(to: CGPoint(x: 3 * sx, y: 4 * sy))
        p.addLine(to: CGPoint(x: 9 * sx, y: 8.5 * sy))
        p.addLine(to: CGPoint(x: 15 * sx, y: 4 * sy))
        // small line at bottom
        p.move(to: CGPoint(x: 6 * sx, y: 13 * sy))
        p.addLine(to: CGPoint(x: 10 * sx, y: 13 * sy))
        return p
    }
}

private struct CampusGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let sx = rect.width / 18.0
        let sy = rect.height / 18.0
        // roof diamond: M9 2L2 6l7 3 7-3-7-4z
        p.move(to: CGPoint(x: 9 * sx, y: 2 * sy))
        p.addLine(to: CGPoint(x: 2 * sx, y: 6 * sy))
        p.addLine(to: CGPoint(x: 9 * sx, y: 9 * sy))
        p.addLine(to: CGPoint(x: 16 * sx, y: 6 * sy))
        p.closeSubpath()
        // curve under
        p.move(to: CGPoint(x: 4 * sx, y: 8 * sy))
        p.addLine(to: CGPoint(x: 4 * sx, y: 11.5 * sy))
        p.addQuadCurve(to: CGPoint(x: 14 * sx, y: 11.5 * sy),
                       control: CGPoint(x: 9 * sx, y: 15.5 * sy))
        p.addLine(to: CGPoint(x: 14 * sx, y: 8 * sy))
        return p
    }
}

File: apps/android/app/src/main/kotlin/dev/forcetower/unes/ui/feature/messages/components/OriginSwatch.kt
package dev.forcetower.unes.ui.feature.messages.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.forcetower.unes.ui.feature.messages.Message
import dev.forcetower.unes.ui.feature.messages.MessageOrigin
import dev.forcetower.unes.ui.feature.messages.MessageOriginMeta
import dev.forcetower.unes.ui.feature.messages.originMeta

// Origin swatch shown at the left of every message row and on the detail
// sender card. Shape and glyph depend on `message.origin`.
//
// * `Direct`     — circle with a tinted ring + person glyph.
// * `Discipline` — rounded rect in the discipline accent with the short code.
// * everything else — rounded tile with a tinted background and a kind icon.
@Composable
internal fun OriginSwatch(
    message: Message,
    modifier: Modifier = Modifier,
    size: Dp = 40.dp,
) {
    val meta = originMeta(message)
    when (message.origin) {
        MessageOrigin.Direct -> DirectSwatch(meta, size, modifier)
        MessageOrigin.Discipline -> DisciplineSwatch(meta, size, modifier)
        else -> IconSwatch(message.origin, meta, size, modifier)
    }
}

@Composable
private fun DirectSwatch(meta: MessageOriginMeta, size: Dp, modifier: Modifier) {
    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape)
            .background(meta.color.copy(alpha = 0.09f))
            .border(1.5.dp, meta.color.copy(alpha = 0.4f), CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Canvas(modifier = Modifier.size(size * 0.38f)) {
            val w = this.size.width
            val h = this.size.height
            val sx = w / 15f
            val sy = h / 15f
            val stroke = Stroke(width = 1.3f * density, cap = StrokeCap.Round, join = StrokeJoin.Round)
            // head: circle radius 2.3 around (7.5, 5)
            drawCircle(
                color = meta.color,
                radius = 2.3f * sx,
                center = androidx.compose.ui.geometry.Offset(7.5f * sx, 5f * sy),
                style = stroke,
            )
            // shoulders curve
            val path = Path().apply {
                moveTo(2.5f * sx, 13f * sy)
                quadraticTo(7.5f * sx, 7.3f * sy, 12.5f * sx, 13f * sy)
            }
            drawPath(path = path, color = meta.color, style = stroke)
        }
    }
}

@Composable
private fun DisciplineSwatch(meta: MessageOriginMeta, size: Dp, modifier: Modifier) {
    Box(
        modifier = modifier
            .size(size)
            .clip(RoundedCornerShape(11.dp))
            .background(meta.color),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = meta.label,
            style = MaterialTheme.typography.labelSmall.copy(
                fontFamily = FontFamily.Monospace,
                fontSize = 10.sp,
                fontWeight = FontWeight.Bold,
                letterSpacing = 0.4.sp,
            ),
            color = Color.White,
        )
    }
}

@Composable
private fun IconSwatch(origin: MessageOrigin, meta: MessageOriginMeta, size: Dp, modifier: Modifier) {
    Box(
        modifier = modifier
            .size(size)
            .clip(RoundedCornerShape(11.dp))
            .background(meta.color.copy(alpha = 0.12f)),
        contentAlignment = Alignment.Center,
    ) {
        OriginGlyph(origin = origin, color = meta.color, size = 18.dp)
    }
}

@Composable
private fun OriginGlyph(origin: MessageOrigin, color: Color, size: Dp) {
    Canvas(modifier = Modifier.size(size)) {
        val w = this.size.width
        val h = this.size.height
        val sx = w / 18f
        val sy = h / 18f
        when (origin) {
            MessageOrigin.App -> {
                // chat card: M4 4h10v7l-4-2-4 3V4z
                val path = Path().apply {
                    moveTo(4f * sx, 4f * sy)
                    lineTo(14f * sx, 4f * sy)
                    lineTo(14f * sx, 11f * sy)
                    lineTo(10f * sx, 9f * sy)
                    lineTo(6f * sx, 12f * sy)
                    close()
                }
                drawPath(
                    path = path,
                    color = color,
                    style = Stroke(width = 1.4f * density, join = StrokeJoin.Round),
                )
            }
            MessageOrigin.Module -> {
                val r = 1.2f
                val side = 5.5f
                val stroke = Stroke(width = 1.4f * density)
                listOf(
                    3f to 3f, 9.5f to 3f, 3f to 9.5f, 9.5f to 9.5f,
                ).forEach { (x, y) ->
                    drawRoundRect(
                        color = color,
                        topLeft = androidx.compose.ui.geometry.Offset(x * sx, y * sy),
                        size = androidx.compose.ui.geometry.Size(side * sx, side * sy),
                        cornerRadius = androidx.compose.ui.geometry.CornerRadius(r * sx, r * sy),
                        style = stroke,
                    )
                }
            }
            MessageOrigin.Secretariat -> {
                val stroke = Stroke(width = 1.3f * density, cap = StrokeCap.Round)
                drawRoundRect(
                    color = color,
                    topLeft = androidx.compose.ui.geometry.Offset(3f * sx, 4f * sy),
                    size = androidx.compose.ui.geometry.Size(12f * sx, 10f * sy),
                    cornerRadius = androidx.compose.ui.geometry.CornerRadius(1.4f * sx, 1.4f * sy),
                    style = stroke,
                )
                // flap M3 4l6 4.5L15 4
                val flap = Path().apply {
                    moveTo(3f * sx, 4f * sy)
                    lineTo(9f * sx, 8.5f * sy)
                    lineTo(15f * sx, 4f * sy)
                }
                drawPath(flap, color = color, style = stroke)
                val mark = Path().apply {
                    moveTo(6f * sx, 13f * sy)
                    lineTo(10f * sx, 13f * sy)
                }
                drawPath(mark, color = color, style = stroke)
            }
            MessageOrigin.Campus -> {
                val stroke = Stroke(width = 1.3f * density, cap = StrokeCap.Round, join = StrokeJoin.Round)
                // roof diamond
                val roof = Path().apply {
                    moveTo(9f * sx, 2f * sy)
                    lineTo(2f * sx, 6f * sy)
                    lineTo(9f * sx, 9f * sy)
                    lineTo(16f * sx, 6f * sy)
                    close()
                }
                drawPath(roof, color = color, style = stroke)
                // cup under the cap
                val cup = Path().apply {
                    moveTo(4f * sx, 8f * sy)
                    lineTo(4f * sx, 11.5f * sy)
                    quadraticTo(9f * sx, 15.5f * sy, 14f * sx, 11.5f * sy)
                    lineTo(14f * sx, 8f * sy)
                }
                drawPath(cup, color = color, style = stroke)
            }
            else -> Unit
        }
    }
}
